File: cs214/projects/project05/quad.cpp
/* quad.cpp computes the roots of a second degree polynomial
 * using the quadratic formula.
 *
 * Input: a, b, c: all doubles;
          root1, root2: the calculated roots as doubles
 * PRE: a, b, and c are real values, root1 and root2 are empty
 * Output: If an error occured the corresponding message is printed
 *         else the roots root1 and root2 are printed.
 *
 * Begun by: Charles Blum, for CS 214 at Calvin College.
 * Completed by: Charles Blum
 * Date: 12 March 2014
 **********************************************************************/

#include <iostream>
#include <cmath>
using namespace std;


bool quadraticRoots(double a, double b, double c,
                        double & root1, double & root2);

int main() {
  cout << "\nFor a quadratic equation (y = ax^2 + bx + c),";
  cout << "\nEnter the roots,";
  cout << "\na=";
  double a, b, c;
  cin >> a;
  cout << "\nb=";
  cin >> b;
  cout << "\nc=";
  cin >> c;

  double root1, root2;
  if ( quadraticRoots(a, b, c, root1, root2) )
  {
  cout << "\nThe first root is " << root1
       << "\n and the second root is " << root2 << endl;    
  }
  else
  {
    cout << "\nAn error occured!  Please see above and try again." << endl;
  }
}

/**************************************************
 * quadraticRoots() computes the roots of a       *
 * second degree polynomial                       *
 * Receive: a, b, c, coefficients for calculation *
 *          root1, root2, destination for         *
 *          calculated roots                      *
 * PRE: a, b, and c are real values               *
 * Passback: root1 - the first root               *
 *           root2 - the second root              *
 **************************************************/
bool quadraticRoots(double a, double b, double c,
                        double & root1, double & root2) {
  if (a != 0)
  {
    double arg = pow(b, 2.0) - 4 * a * c;
    if (arg >= 0)
    {
      root1 = (-b + sqrt(arg))/(2*a);
      root2 = (-b - sqrt(arg))/(2*a);
      return true;
    }
    else 
    {
      cerr << "\n*** quadraticRoots(): b^2 - 4ac is negative!" << endl;
      root1 = root2 = 0.0;
      return false;
    }
  }
  else
  {
    cerr << "\n*** QuadraticRoots(): a is zero!" << endl;
    root1 = root2 = 0.0;
    return false;
  }
}

File: cs214/projects/project05/quad.rb
# quad.rb computes the roots of a 2nd degree polynomial
# using the quadratic formula
#
# Input: a, b, c, the coefficients of the polynomial.
# Precondition: a, b, and c are real numbers.
# Output: error message accordingly if coefficients don't
#         return a real number, other wise root1 and root2.
#
# Begun by: Charles Blum, for CS 214 at Calvin College.
# Completed by:Charles Blum
# Date: 12 March 2014
##################################################################


# quadraticRoots() computes the roots using the quadratic formula
# Receive: a, b, c, the polynomial coefficients,  
# PRE: a, b, c are all real numbers.
# Return: root1, root2, the calculated roots, 
#         valid, determines if an error occured during calculation
##################################################################

def quadRoots( a, b, c )
  #form of roots: [ validity (boolean), root1 (float), root2 (float) ]
  if a == 0
    puts "Coefficient a is Zero! Please try again.";
    roots = [ false, 0.0, 0.0]
  else
    arg = ((b ** 2) - (4 * a * c))
    if arg < 0
      puts "b^2 - 4ac is negative! Please Try again.";
      roots = [ false, 0.0, 0.0 ]
    else
      roots = [ true, ((-b + Math.sqrt(arg))/(2 * a)), ((-b - Math.sqrt(arg))/(2 * a)) ]
    end
  end
end 

def main
  puts "Calculate the roots of a 2nd degree polymonial"; 
  puts "( y = ax^2 + bx + c ) using the quadratic formula:";
  print "a=";
  a = gets.to_f
  print "b="; 
  b = gets.to_f
  print "c="; 
  c = gets.to_f
  roots = quadRoots( a, b, c )
  if roots[0]
    puts "The first root is: #{roots[1]}";
    puts "and the second root is: #{roots[2]}";
  end
end

main
